Docs and improved components

fadeev opened the pull request:
Updated Storybook and simplified `main.js` and `preview.js`. Documenting components using jsdoc-style comments. Also, modified API of Footer component to accept multiple props rather than one object—this makes it easier to document props and add defaults.

I think using `jsdoc` is a good idea, because the documentation becomes part of components, rather than something we use Storybook for. Even if we stop using Storybook, components will still be properly documented.

Not using tables for CSS variables, because it's a pain to edit them (it's no Emacs org-mode ;-)

https://github.com/allinbits/design/issues/197

That should go without saying, but feel free push doc updates to this branch.

closes https://github.com/allinbits/design/issues/211

fadeev commented:
Added sheet transition from the bottom in the sidebar, preliminary modal as part of the sidebar. Customisable card component with read more and no overlapping shadows (if multiple cards are close, shadows stay behind).

nassdonald commented:
This is awesome progress. I’m just playing around and trying to break the sidebar, got this from opening and closing repeatedly:

```
undefined is not an object (evaluating 'this.$refs.sidebar.style')
touchend@https://deploy-preview-133--cosmos-ui.netlify.com/main.f6d1b6dd28beb4addc8b.bundle.js:1:41745
touchend@[native code]
invokeWithErrorHandling@https://deploy-preview-133--cosmos-ui.netlify.com/vendors~main.f6d1b6dd28beb4addc8b.bundle.js:2:416126
```

nassdonald commented:
There also seems to be a bug with the modal positioning on mobile. The top of the modal animates up beyond the top of the viewport, requiring the user to scroll up to see the very top. This also causes the close button to reposition when the transition is completed. Can provide screen recording if needed?

fadeev commented:
Thanks, Nass!

> undefined is not an object (evaluating 'this.$refs.sidebar.style')

That's helpful! I'll add conditionals to test for `undefined`.

The modal issue, on the other hand, is strange. Are you talking about `Modal` component or `Sidebar` with `side: center`? Because the new modal is implemented in the Sidebar component and as of now it doesn't have a close button at all (unless you're talking about "Open sidebar" button visible below the scrim). In any case, the more data I have, the more bugs I can fix!

fadeev commented:
Added `CodeBlock` component to the repo. Fixed top and bottom padding, moved scrollbar to the bottom.

<img width="503" alt="Screenshot 2020-02-15 at 22 20 36" src="https://user-images.githubusercontent.com/332151/74592414-0bd8a600-5043-11ea-85cc-1362ee05f3a1.png">

<img width="876" alt="Screenshot 2020-02-15 at 22 23 49" src="https://user-images.githubusercontent.com/332151/74592415-0e3b0000-5043-11ea-8cba-c2043479eb01.png">

nassdonald commented:
> Are you talking about Modal component or Sidebar with side: center?

Ah yes I was looking at the old modal component, my bad. The new one doesn't have this problem since it doesn't go full-screen on mobile and doesn't have a close button.

On that note, perhaps that should be one of the props for the component? Modals go full-screen on handsets unless specified otherwise by a prop: `fullScreenHandset: true(default) | false`. I'm sure you'll come up with a better name. 😄 

Also, I think you've got the right idea with not including the close button as part of the component, since it will depend on the content and UI of the modal itself when building a close button. 👍

nassdonald commented on src/CodeBlock/CodeBlock.stories.js:
```diff
@@ -0,0 +1,27 @@
+      <code-block :value="data.short"/>
+      <p>Multiline snippet with syntax highlighting:</p>
+      <code-block :value="data.medium" language="go"/>
+      <p>Multiline snippet with syntax highlighting and expand button:</p>
```
> …and external file link.

What should we call that actually?

fadeev replied:
Can't seem to find the source "and external file link" quote. What do you mean?

fadeev commented:
@nassdonald I'm actually thinking about adding a close button for `center` modal, so that it can be used without additional config, with an option to disable the button.

nassdonald commented:
Yeah the bottom modal sheet is much better now. Though I believe the best practise is for modals to scroll with the body, rather than their own internal scrolling containers. I can illustrate this better with a codepen or a sketch or something. Not crucial, but would be better UX and accessibility. Vertical scrolling in divs is generally best avoided on mobile, if possible.

Also, an option to scroll up to close would be cool but might be tricky to implement - not sure if you can calculate negative scrollTop values. Also not crucial.

I’ll have a think about a suitable threshold for the amount of scrim visible on the bottom sheet. Do we have a way to test when the bottom sheet has less content? Would be good to stress test all components with no content, little content, and lots of content.

As for the horizontal swiping issue, it seems to be fine when swiping from the scrim, but not when swiping from the sheet itself.

I’ll upload a video to illustrate the behavior on mobile. Bear with me...

fadeev commented:
Thanks, Nass! I'm sorry, if it takes a while to iron out these issues. I wish I could write tests for this, but there are literally no good solutions for testing such interactions.

I think I've fixed the horizontal swiping issue.

fadeev commented:
I'm not sure the body should be scrolled with the bottom sheet. The sheet transitions in by sliding over the body: the body stays still in the background. I've used Slack for Android as a reference: https://imgur.com/a/lVJXtMl

I've used "random" channel as a background as to not share any info, this is a public repo, after all 😄

nassdonald commented:
> there are literally no good solutions for testing such interactions.

I think these things will always require manual human testing. It's got to "feel" right, and machines can't really ever automate testing that kind of thing.

> I think I've fixed the horizontal swiping issue.

Testing now…

nassdonald commented:
Okay it's **almost** there… currently this is what happens

1. Swipe left on sheet
2. Sheet moves left
3. Scroll up and down while swiping
4. Vertical scrolling occurs on sheet but sheet stops moving horizontally
5. Release touch
6. Sheet returns to original position

This is an improvement but not completely fixed. I wouldn't break your neck over it. But if it's a simple fix, disabling vertical scrolling on touch would be ideal.

nassdonald commented:
> The sheet transitions in by sliding over the body: the body stays still in the background.

When I say "body" I don't mean the main content behind, I mean the `<body>` component. I've mocked up a [rough example](https://codepen.io/nassdonald/pen/axryRV) of this to show you what I mean. This works by saving the scroll position before the modal opens and restoring it when closed. It's not bulletproof, but it does allow us to use the body for all scrolling, which is more user-friendly (and mobile-friendly) than a scrolling div.

I might be wrong about this and this could be a can of worms, but I think we should strive to get the best modal behavior possible, even if it means challenging the status quo. 🤓

fadeev commented:
Modified Stars component to be more like the background from the video. Since this is just a canvas, simple zooming can be achieved by scaling the parent container. Tried it on hover, works like a charm. Didn't pay much attention to the background, because it can be any image or gradient (canvas is transparent).

<img width="1158" alt="Screenshot 2020-02-21 at 15 05 01" src="https://user-images.githubusercontent.com/332151/75025041-48315980-54bc-11ea-9fee-bb27c8e72d85.png">

nassdonald commented:
This is great @fadeev! The only adjustment I'd make with the movements is restricting the stars to one direction. In the video, they never change direction, they simply appear and disappear in a straight path. Currently I believe there are a few stars in your version that seem to follow the sine path you were talking about.

fadeev commented:
Oh, I thought their trajectory was what we needed. Ok, I'll turn if off.

nassdonald commented:
> Oh, I thought their trajectory was what we needed.

Yeah that's what I'm saying. In the video, the stars never change direction. This is what I'm seeing:

![image](https://user-images.githubusercontent.com/2961244/75039554-d6074780-54b0-11ea-8c77-636120b62d1d.png)

fadeev commented:
Fixed the issue above, added knobs 🎚  Look! It's gliiter! 🦄

![DLtibME - Imgur](https://user-images.githubusercontent.com/332151/75046383-c86cb480-54e6-11ea-8dbe-2d1ce9a9beab.gif)

fadeev commented:
It's insane. I've spent hours trying to figure out why sometimes sidebar's overlay allows scrolling on iOS and sometimes doesn't. Turns out, it depends on *whether the URL bar is visible* at the moment you open the sidebar. If it is not, `overflow: hidden` has no effect until you scroll back URL bar into view (then it works). If it is, everything works fine.

Also, I wanted center modal to both transition in with scale transform and position itself correctly based on content height. Had to inspect diffs from the last ten commits until I realized the problem was there all along: `el.getBoundingClientRect().height` and `el.offsetHeight` return different results when an element is being transitioned.

nassdonald commented:
Ach that's no fun! Sorry you've had such a battle with iOS. 😞 Well if it's any consolation, the body is not scrolling in either case while the modal is visible on my iPhone SE. Works as intended for me!  🎉 Maybe it's a device-specific bug? Or certain versions of Mobile Safari? 🤷‍♂ 

I'd say there's only one edge case issue left. When the center modal is shown on mobile, while the URL bar is **not** visible, then the URL bar is expanded (by tapping on it), the modal no longer scrolls all the way to the bottom. There is some content hidden behind the bottom toolbar, most likely since the viewport resize isn't triggering a recalculation of modal height?

---

A few extra thoughts…

How do we determine whether or not a modal goes _fullscreen_ on mobile? Should be a prop, I think. For example, there are cases for making a center modal fullscreen on mobile (if the content overflows), but there are also cases for having a simple, typical dialog style (centered with intrinsic height) on mobile:
![image](https://user-images.githubusercontent.com/2961244/75260184-221ff800-57e1-11ea-9919-7e24eb2eb628.png)

I also wonder if we can change the animation for _fullscreen_ center modals on mobile to slide up instead of fade in with scaling ("pop in") transition? This would be more familiar to mobile users, since it appears like a typical fullscreen modal dialog.

fadeev commented:
Yes, fullscreen is an open question for now. I'm thinking we can have the following behaviour: the width is determined by `width` and `maxWidth` props (or defaults). Let's say `width="500px" maxWidth="90%"`. The modal will stay centered until viewport goes below 500px, at which point it will go fullscreen, unless `fullscreen="false"` is set (like, for a dialog box). I think that should cover all use-cases.

fadeev commented:
Tailwindin' ⛵️

https://deploy-preview-133--cosmos-ui.netlify.com/iframe.html?id=sidebar--center

fadeev commented:
Links in the Footer's story were messy for the purpose of showing that you can pass different kinds of objects (with or without title or icon). Should've mentioned that in a comment. Not a big deal, though, let's keep it tidy.

lovincyrus commented:
> Links in the Footer's story were messy for the purpose of showing that you can pass different kinds of objects (with or without title or icon). Should've mentioned that in a comment. Not a big deal, though, let's keep it tidy.

Makes sense.

fadeev commented:
There is at least opne thing we should do before we proceed with merging this. We use `Footer`, `Modal` and `Tooltip` on cosmos.network. Since what is known atm as `Sidebar` superseeds `Modal` we should rename `Sidebar` to `Modal` and update it, so that it works with cosmos.network.

fadeev commented:
This branch and https://github.com/cosmos/cosmos.network/pull/765 should be working fine. I've added a temporary close button, so that it would be possible to close the modal on small screens. Also, deleted old modal and renamed `Sidebar` to `Modal`.

lovincyrus commented on src/Card/Card.vue:
```diff
@@ -0,0 +1,241 @@
+          <!-- @slot Main content of the card -->
+          <slot/>
+          <transition name="fade">
+            <div class="body__more" @click="expand" v-if="canExpand && !isExpanded">read more</div>
```
![image](https://user-images.githubusercontent.com/1021101/75818422-b217f180-5d4d-11ea-9178-5685782838f1.png)

Not sure how we display `read more` is the conventional way 🤔 @fadeev

fadeev replied:
Well, we're not using this component and probably won't be it at all, so it doesn't matter.

fadeev commented:
This is how it looks for me. Not the purtiest close button, but I guess it's ok for now, tried adding circle behind and it felt too heavy. So, yeah, I'm merging.

<img width="1447" alt="Screenshot 2020-03-04 at 10 48 27" src="https://user-images.githubusercontent.com/332151/75849410-97ee1a00-5e06-11ea-9e1d-1a895a13121e.png">

<img width="816" alt="Screenshot 2020-03-04 at 10 50 02" src="https://user-images.githubusercontent.com/332151/75849424-9cb2ce00-5e06-11ea-9996-5826358b2709.png">

<img width="474" alt="Screenshot 2020-03-04 at 10 50 08" src="https://user-images.githubusercontent.com/332151/75849427-9f152800-5e06-11ea-931d-35a756c3819b.png">
